compilers/concrete-compiler/compiler/lib/Dialect/RT/Transforms/HoistAwaitFuturePass.cpp

```diff
@@ -0,0 +1,259 @@
+// Part of the Concrete Compiler Project, under the BSD3 License with Zama
+// Exceptions. See
+// https://github.com/zama-ai/concrete-compiler-internal/blob/main/LICENSE.txt
+// for license information.
+
+#include <concretelang/Analysis/StaticLoops.h>
+#include <concretelang/Dialect/RT/IR/RTDialect.h>
+#include <concretelang/Dialect/RT/IR/RTOps.h>
+#include <concretelang/Dialect/RT/Transforms/Passes.h>
+
+#include <mlir/Dialect/Utils/StaticValueUtils.h>
+
+#include <iterator>
+#include <optional>
+
+namespace {
+struct HoistAwaitFuturePass
+    : public HoistAwaitFuturePassBase<HoistAwaitFuturePass> {
+  // Checks if all values of `a` are sizes of a non-dynamic dimensions
+  bool allStatic(llvm::ArrayRef<int64_t> a) {
+    return llvm::all_of(
+        a, [](int64_t r) { return !mlir::ShapedType::isDynamic(r); });
+  }
+
+  void runOnOperation() override {
+    mlir::func::FuncOp func = getOperation();
+
+    llvm::SmallVector<mlir::Operation *> opsToErase;
+
+    func.walk([&](mlir::concretelang::RT::AwaitFutureOp awaitFutureOp) {
+      // Make sure there are no other consumers that rely on the
+      // synchronization
+      if (!awaitFutureOp.getResult().hasOneUse())
+        return;
```

Comment:
This restriction feels a bit too strong - is it not enough if the uses are dominated by the new location of the awaitFutureOp? Not essential for this PR now, but we may lose some cases.

Reply:
This is indeed a bit conservative. Let's have a first iteration and then relax the constraints.